### PR TITLE
Fix/992 notify data file search selection changes

### LINKED ISSUE #992: As an OFA admin/STT user, I want to be notified if I change search selections after a file has been uploaded but not yet submitted
Expected behavior:

1. make initial selections
2. upload file (but do not submit)
3. change selections in combo-box
4. click search
5. modal pop up saying: ` (H1) Files have not been submitted. (body) Searching without submitting will discard your changes and remove any uploaded files`
6. Button updated to `Cancel` and `Discard files and search`

Acceptance Criteria:
- [ ] Frontend raises [an error/warning notification](https://www.figma.com/file/irgQPLTrajxCXNiYBTEnMV/TDP-Mockups-For-Feedback?node-id=4203%3A22618) if user changes selections after upload but prior to submit
- [ ] Frontend matches mock ups

Tasks:
- [x] Meet w/ UX team to discuss list of ACs (or invite to dev sync)
- [ ] Implement changes as proposed by UX and @ADPennington 

Notes:
- Related [finding](https://github.com/raft-tech/TANF-app/pull/787#issuecomment-829591282) re: ensuring data uploaded is associated with users final selections 
- For future: need a diff ticket for trying to navigate away from page. 
- more discussion needed around submission history epic

### PR DESCRIPTION
**Pull request by jtimpe**
## Summary of Changes
Pull request closes #992

1. Notify the user when they are about to change their search params without having submitted their attached file.
2. ~~Change the fiscal year header on the reporting page to update without the search button being clicked (related finding mentioned in ticket details)~~

## How to Test
```
cd tdrs-frontend && npm run start
cd tdrs-backend && docker-compose up
```

### 1. Notify the user when they are about to change their search params without having submitted their attached file.
1. Open http://localhost:3000/ and sign in.
2. Approve your user and assign a role. In the frontend, navigate to the "Data Files" tab. Apply some search filters and click "search".
3. Attach a data file, but do not submit it.
4. Change the search filters, click "Search". A modal should display notifying the user that their attached file will be cleared if they continue with the search. they have two options:
   - a. Cancel - this resets the search filters to _their previous values_ (before the user clicked "Search") and maintains the attached file
   - b. Search - this clears the attached file and proceeds with the search, fetching files for the newly selected parameters

### ~~2. Change the fiscal year header on the reporting page to update without the search button being clicked.~~
1. ~~Open http://localhost:3000/ and sign in.~~
2. ~~Approve your user and assign a role. In the frontend, navigate to the "Data Files" tab. Apply some search filters and click "search".~~
3. ~~Change the search filter values. The report header in the middle of the page should "live update" without the need for clicking the search button.~~

This functionality was removed in favor of a more accessibility-friendly page flow. The user is now required to actually submit the search input form in order to both refresh the search screen and set the global state values for data file submission.

## Deliverables
_More details on how deliverables herein are assessed included [here](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] Frontend raises [an error/warning notification](https://www.figma.com/file/irgQPLTrajxCXNiYBTEnMV/TDP-Mockups-For-Feedback?node-id=4203%3A22618) if user changes selections after upload but prior to submit
+ [x] Frontend matches mock ups
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [x] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Comment by andrew-jameson** [comment]
Negative case #1: 
> data-files/ as Developer
> Search 2022/qtr1
> select file section 1
> select quarter 2, don't click search
> banner updates to qtr 2
> select file section 2
> click submit, success
> check backend LogEntry, noted as Q2 for both files

Negative case #2: 
> data-files/ as Developer
> Search 2022/qtr3
> select file X section 1
> submit
> select quarter 4, don't click search
> banner updates to qtr 2
> select file Y for section 2
> submit
>backend notes files as:
fileX quarter 3
file Y quarter 4
> does not submit fileX despite quarter 4 being selected



cc @ADPennington @reitermb  Can we talk about this behavior? I feel like the error alert should force the user to reclick search.

File: tdrs-frontend/src/components/Reports/Reports.jsx [hunk 1]
```diff
@@ -280,18 +295,53 @@ function Reports() {
...
         <UploadReport
           stt={stt?.id}
-          header={submittedHeader}
+          header={`${currentStt} - Fiscal Year ${selectedYear} - ${quarters[selectedQuarter]}`}
           handleCancel={() => setIsToggled(false)}
         />
       )}
+      <Modal
+        title="Files Not Submitted"
+        message="Your files have not been submitted. Searching without submitting will discard your changes and remove any uploaded"
```
Comment (raftmsohani) [inline tdrs-frontend/src/components/Reports/Reports.jsx]: Is the message complete? Maybe "...Uploaded FILE"?

**Review by raftmsohani — COMMENTED** [review]
LGTM! Tested locally

File: tdrs-frontend/src/components/Reports/Reports.jsx (hunk 1 above)
Comment (reitermb) [reply tdrs-frontend/src/components/Reports/Reports.jsx]: Nice catch!  Probably uploaded "file(s)" to also include the possibility of multiple non-submitted files.

Comment (jtimpe) [reply tdrs-frontend/src/components/Reports/Reports.jsx]: wow, not sure how that happened honestly. code goblins 😄

**Comment by jtimpe** [comment]
per @reitermb in MM, documented for posterity
> If we go for the disabled state (with the appropriate aria) that should do the trick from an a11y compliance perspective. 

> That said, my preference would still be ignoring form changes until 'Search' gets clicked rather than auto-updating anything below the search form—in this case including the button.

> Ex. of what would be allowed: 
Search for Q2 2022
Selecting a file
Changing the quarter field to Q3 & NOT clicking search,
Clicking submit
Selected file gets uploaded to Q2 2022 (e.g. what the heading below search still reads as)

File: tdrs-frontend/src/components/Reports/Reports.test.js [hunk 2]
```diff
@@ -312,13 +319,16 @@ describe('Reports', () => {
...
 
     // There should be 4 more dispatches upon making the submission,
     // one request to /reports for each file
+    await waitFor(() =>
+      expect(getByText('Submit Data Files')).toBeInTheDocument()
+    )
     fireEvent.click(getByText('Submit Data Files'))
     await waitFor(() => getByRole('alert'))
-    expect(store.dispatch).toHaveBeenCalledTimes(14)
+    expect(store.dispatch).toHaveBeenCalledTimes(18)
```
Comment (andrew-jameson) [inline tdrs-frontend/src/components/Reports/Reports.test.js:333]: Why did these change? Seems odd

Comment (jtimpe) [reply tdrs-frontend/src/components/Reports/Reports.test.js:333]: I changed the order of event dispatches in order to be able to "reset" the input values without changing the redux state. basically for the tests that means there's one extra dispatch for each search input after the search button is clicked (before, the values were already in redux state, so the update didn't need to be dispatched).

**Review by andrew-jameson — APPROVED** [review]

**Review by reitermb — CHANGES_REQUESTED** [review]
Running into similar a11y issues here as when we first added the session timeout modal. The modal doesn't grab or trap focus or get announced when it appears and the "dimmed" background seems to be below the file pickers in z-index. 
image

**Review by raftmsohani — APPROVED** [review]
LGTM!

File: tdrs-frontend/src/assets/Modal.scss [hunk 3]
```diff
@@ -1,13 +1,13 @@
 .modal {
-  position: fixed; 
-  z-index: 1; 
+  position: fixed;
+  z-index: 3;
```
Comment (jtimpe) [inline tdrs-frontend/src/assets/Modal.scss:3]: the actual change to fix the highlight effect when the modal was displayed. the rest are style changes made by my editor

File: tdrs-frontend/src/components/Reports/Reports.jsx [hunk 4]
```diff
@@ -280,18 +300,55 @@ function Reports() {
...
       {isUploadReportToggled && (
         <UploadReport
           stt={stt?.id}
-          header={submittedHeader}
-          handleCancel={() => setIsToggled(false)}
+          header={`${currentStt} - Fiscal Year ${selectedYear} - ${quarters[selectedQuarter]}`}
+          handleCancel={() => {
+            setIsToggled(false)
+            resetPreviousValues()
+            dispatch(clearFileList())
```
Comment (jtimpe) [inline tdrs-frontend/src/components/Reports/Reports.jsx:325]: make sure files are cleared on cancel so the modal doesn't display if the user tries to execute a new search

File: tdrs-frontend/src/components/Modal/Modal.jsx [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    () =>
+      isVisible && modalRef && modalRef.current
+        ? modalRef.current
+            .querySelector(`button[buttonkey="${buttons[0].key}"]`)
+            .focus()
+        : undefined,
+    [isVisible, modalRef, buttons]
+  )
+
+  const onTabPressed = (shiftKey = false) => {
```
Comment (jtimpe) [inline tdrs-frontend/src/components/Modal/Modal.jsx:14]: along with the `focus()` in the `useEffect` above, this is what "traps" the keyboard focus into the modal. the user should only be able to tab through (forward or backward by adding a `shiftKey` modifier`) buttons within the modal's list.

**Review by reitermb — CHANGES_REQUESTED** [review]
Looking good on focus trapping and visual z-index issues!

One last thing it still needs is an initial focus set to the modal's H1 though. 
Currently the screenreader behavior results in a lack of context as to what the modal buttons are referring to. 

e.g. 
*Modal opens*
Screenreader says "Cancel"

The desired behavior here would be for focus to land on that H1 "Files Not Submitted" heading so that the behavior is:
*Modal Opens*
Screenreader says "Files Not Submitted"
User can then tab to buttons and have context for what they refer to

---

If it's not scope creeping this ticket it would be great to also give the paragraph text an ID that can then be referenced in an aria-describeby property on the modal's h1. 

This would result in the screenreader user also being read the paragraph text, not just the heading text and enhance the overall clarity of the modal. That said, we can make this a follow-on issue if this adds scope (noting that this could also be an enhancement to our idle timer modal).

**Comment by jtimpe** [comment]
@reitermb thanks! i feel like we're getting close, pushed up both the requested changes. i tested the header focus change and that seems to work, but i don't have a screen reader set up to test the `aria-describedby` attribute. is there something you recommend?

also, thinking ahead based on the convo in tabletop, i think it'd be helpful to discuss more about these a11y rules in general so we can make sure we're designing these reusable components appropriately and planning that into the initial scope of the work.

**Comment by ADPennington** [comment]
@jtimpe @reitermb @ttran-hub is this ready for my review?

**Review by ADPennington — APPROVED** [review]
👍🏾 confirmed the following:

- changing selections as a `data analyst` or `ofa admin` after upload and clicking search triggers the warning modal
- clicking `cancel` in the modal returns the user to the same changed selections that do not match the original selections reflected in the header description ✔️ 
- clicking `discard and search` in the modal, discards the uploaded files and shows the user to the results of the new search. ✔️ 
- if i change selections without clicking search and submit, the file submitted will have metadata that matches the header (the original selections ✔️ 

this is qasp- approved pending `gov a11y review` cc: @ttran-hub @reitermb

**Comment by ADPennington** [comment]
> 👍🏾 confirmed the following:
> 
> * changing selections as a `data analyst` or `ofa admin` after upload and clicking search triggers the warning modal
> * clicking `cancel` in the modal returns the user to the same changed selections that do not match the original selections reflected in the header description ✔️
> * clicking `discard and search` in the modal, discards the uploaded files and shows the user to the results of the new search. ✔️
> * if i change selections without clicking search and submit, the file submitted will have metadata that matches the header (the original selections ✔️
> 
> this is qasp- approved pending `gov a11y review` cc: @ttran-hub @reitermb

@reitermb @ttran-hub this PR is in the [qasp environment](https://tdp-frontend-qasp.app.cloud.gov/data-files), which is only accessibly using non-acf credentials for sign-in.

**Review by reitermb — APPROVED** [review]
Working great! That should do it for everything @ttran-hub and I discussed, including the describeby stretch goal. 💯 

Video of it in action:
Modal pops up
Focus reads out heading
Paragraph text gets read automatically after heading
Maintains existing focus trap to contain users to the modal

https://user-images.githubusercontent.com/28515874/199578147-3cce0222-e3b6-4c45-86be-268f0fc96578.mp4

cc: @ADPennington

**Comment by ADPennington** [comment]
> Working great! That should do it for everything @ttran-hub and I discussed, including the describeby stretch goal. 💯
> 
> Video of it in action: Modal pops up Focus reads out heading Paragraph text gets read automatically after heading Maintains existing focus trap to contain users to the modal
> 
>  chrome_5HxmtE6VY4.mp4 
> cc: @ADPennington

thanks @reitermb and @ttran-hub!  @andrew-jameson this is ready to merge.

**Comment by reitermb** [comment]
> @reitermb thanks! i feel like we're getting close, pushed up both the requested changes. i tested the header focus change and that seems to work, but i don't have a screen reader set up to test the `aria-describedby` attribute. is there something you recommend?
> 
> also, thinking ahead based on the convo in tabletop, i think it'd be helpful to discuss more about these a11y rules in general so we can make sure we're designing these reusable components appropriately and planning that into the initial scope of the work.

👍 Always happy to talk a11y—is there a first step kind of sync you think would be helpful there?

Re: Screenreaders, Thomas generally tests with NVDA—a popular screenreader for Windows. My main testing is a combination of that and Voiceover (for MacOS/iOS/iPadOS) with occasional Talkback (Android) testing thrown in just revisit mobile occasionally. In theory Voiceover is going to behave most consistently paired with Safari and NVDA paired with Firefox—but in practice we get consistent behavior cross-browser and cross-platform. 

Pending further syncing I also have a hack.md with some broader commentary on a11y tools that might be useful: https://hackmd.io/Z5t247a7St2tZzLS35JBkw?view

**Pull request merged**